OrbitCore/FunctionFramepointerValidator.cpp

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "FunctionFramepointerValidator.h"
+
+#include <capstone/capstone.h>
+#include <iostream>
+
+FunctionFramepointerValidator::FunctionFramepointerValidator(
+    csh handle, const uint8_t* code, size_t code_size) {
+  handle_ = handle;
+  instructions_ = nullptr;
+  instructions_count_ =
+      cs_disasm(handle, code, code_size, 0, 0, &instructions_);
+}
+
+FunctionFramepointerValidator::~FunctionFramepointerValidator() {
+  cs_free(instructions_, instructions_count_);
+}
+
+bool FunctionFramepointerValidator::IsCallInstruction(
+    const cs_insn& instruction) {
+  for (uint8_t i = 0; i < instruction.detail->groups_count; i++) {
+    if (instruction.detail->groups[i] == X86_GRP_CALL) return true;
+  }
+  return false;
+}
+
+bool FunctionFramepointerValidator::IsMovInstruction(
+    const cs_insn& instruction) {
+  return instruction.id == X86_INS_MOV || instruction.id == X86_INS_MOVQ;
+}
+
+bool FunctionFramepointerValidator::IsBasePointer(uint16_t reg) {
+  return reg == X86_REG_BP || reg == X86_REG_EBP || reg == X86_REG_RBP;
+}
+
+bool FunctionFramepointerValidator::IsStackPointer(uint16_t reg) {
+  return reg == X86_REG_SP || reg == X86_REG_ESP || reg == X86_REG_RSP;
+}
+
+bool FunctionFramepointerValidator::ValidatePrologue() {
+  cs_regs regs_read, regs_write;
+  uint8_t read_count, write_count;
+
+  // check the first instruction: it must be "push ebp"
+  if (cs_regs_access(handle_, &instructions_[0], regs_read, &read_count,
+                     regs_write, &write_count) == 0) {
+    if (instructions_[0].id != X86_INS_PUSH || read_count != 2 ||
+        !IsStackPointer(regs_read[0]) || !IsBasePointer(regs_read[1])) {
+      return false;
+    }
+  }
+
+  // check the second instruction: it must be "mov ebp, esp"
+  if (cs_regs_access(handle_, &instructions_[1], regs_read, &read_count,
+                     regs_write, &write_count) == 0) {
+    if (!IsMovInstruction(instructions_[1]) || read_count != 1 ||
+        !IsStackPointer(regs_read[0]) || write_count != 1 ||
+        !IsBasePointer(regs_write[0])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// We will actually only check, whether there is any correct epilogue.
+// It might be the case, that there are multiple function returns and that not
+// all are correct. However, we would not expect a compiler to produce this,
```

Comment:
I'm pretty sure compilers produce multiple ret instructions in optimized code (maybe not together with "mov esp, ebp" and "pop ebp", but for the sake of this comment...)